### PR TITLE
Add sampled logging when workflow query task times out

### PR DESCRIPTION
**Pull request by gow**
## What changed?
Adding a sampled logging when a workflow query task times out.

## Why?
Need this to be able to debug issues in the future when workflow queries timeout. 

## How did you test it?

Ran it locally with the sampling rate set to 1 and simulated a timeout. Observed that the data is logged.

## Potential risks

N/A

## Is hotfix candidate?

No

### CODE REVIEW

File: service/matching/matching_engine.go
Comment (gow) [inline service/matching/matching_engine.go:878]: How can we get the workflow ID here? It seems to be useful to be logged.

File: service/matching/matching_engine.go
Comment (dnr) [inline service/matching/matching_engine.go:877]: We already have a `throttledLogger` available in most places.. are you sure we shouldn't use that?

File: service/matching/matching_engine.go
Comment (dnr) [inline service/matching/matching_engine.go:869]: Instead of calling namespace registry, constructing log tags, etc. yet again, we have a `tqm` here, which already has a tagged logger and a `taskQueueConfig` with tq-specific configs. Can we somehow delegate this logging to that, or do this inside of that?

File: service/matching/config.go
Comment (dnr) [inline service/matching/config.go:224]: Why not filtered by task queue also?

File: service/matching/matching_engine.go
Comment (dnr) [reply service/matching/matching_engine.go:878]: It's in the request (in the embedded request, in the execution)

File: service/matching/config.go
Comment (gow) [reply service/matching/config.go:224]: Not completely sure. But from reading the description in the task, we actually don't know which workflowID and task queue for the customer were troublesome. So, I'm guessing we can't filter this config by task queue (?). The requirement is to log the workflow ID and the task queue so that we can troubleshoot easily.
@yycptt thoughts?

File: service/matching/matching_engine.go
Comment (gow) [reply service/matching/matching_engine.go:869]: I couldn't find a tagged logger in taskQueueManager implementation [here](https://github.com/temporalio/temporal/blob/chetan_query_workflow_logging/service/matching/task_queue_manager.go#L162-L195). There is a `taggedMetricsHandler` inside it but no tagged logger. Did I miss something?

File: service/matching/matching_engine.go
Comment (gow) [reply service/matching/matching_engine.go:877]: Discussed this with @yycptt . The concern is that the `throttledLogger` is shared. If we turn this on for a customer, it could spew a lot of log lines and consume full quota and other parts of matching engine logging may get throttled. Similarly if some other part triggers throttling, we may not see the log lines we expected to inspect.

File: service/matching/matching_engine.go
Comment (gow) [reply service/matching/matching_engine.go:878]: Done.

File: service/matching/matching_engine.go
Comment (dnr) [reply service/matching/matching_engine.go:869]: That logger is tagged, see https://github.com/temporalio/temporal/blob/chetan_query_workflow_logging/service/matching/task_queue_manager.go#L235-L238
(There's no separate type for it so we just call it a logger. Arguably the metricsHandler should just be called that and not have "tagged" in the name either.)

File: service/matching/matching_engine.go
Comment (dnr) [reply service/matching/matching_engine.go:877]: Sure, that's fair.

I've always been vaguely unhappy with the throttled logger anyway. It seems like what we want is a sampling+auto-adjusting rate limited logger: you construct it (at the top level) with a max rate, all calls pass a sample rate, and by default it samples at that rate. If the overall rate exceeds the max (using exponential average), it automatically adjusts all sample rates downward to stay below the overall max. Or maybe build in some fairness and have it bias toward keeping logs from less frequent call sites. Of course this is future work, this PR doesn't have to be blocked on that.

File: config/dynamicconfig/development-sql.yaml
Comment (gow) [reply config/dynamicconfig/development-sql.yaml:52]: Not sure what you mean by keeping them consistent. Did you mean group them by name? There are no `matching.xyz` configs in this file. So I just put it in the end.

File: service/matching/config.go
Comment (gow) [reply service/matching/config.go:224]: ok. Sounds good. So what you're saying is it's harmless to filter by ns+task_queue since you can always skip specifying task queue in the configs?

File: service/matching/matching_engine.go
Comment (gow) [reply service/matching/matching_engine.go:869]: Got it. I'll see how I can log using `tqm`. It's often not a good idea to reach into internals of other objects. So, I'll probably make the matcher's logger be auto tagged as well if it's easier.
As far as logging location, this seems like the right location since we are handling the context deadline and is at a reasonably high level to log the data.

File: config/dynamicconfig/development-sql.yaml
Comment (dnr) [reply config/dynamicconfig/development-sql.yaml:52]: I mean making the parallel change to `development-cass.yaml` so the behavior doesn't depend on which db you're testing with

File: service/matching/config.go
Comment (dnr) [reply service/matching/config.go:224]: exactly. filtering by ns+task queue allows any of: ns+tq+tqtype, ns+tq, tq (shouldn't be used), ns, global.
see https://github.com/temporalio/temporal/blob/main/common/dynamicconfig/collection.go#L510-L521

File: service/matching/matching_engine.go
Comment (dnr) [reply service/matching/matching_engine.go:869]: Yeah, I see that it may be awkward to do, I was just noticing the duplication and want to see if there's a way to refactor it. E.g. possibly moving everything after the call to `tqm.DispatchQueryTask` inside of that function or in another function on the tqm (including the query map). That's a larger change and may be bad for other reasons, but worth thinking about.

Comment (gow) [reply service/matching/matching_engine.go:869]: Yes. Looks like we can't have a tagged logger in matcher since its internal state doesn't have the namespace, queue name/type etc like `tqm`. So a better approach is to rethink this whole code block and see if it belongs inside TaskQueueManager.
Even that doesn't seem like the right place to me since the channel is created here and attached to `e.queryResults` and eventually read here. So a refactor would lead to a larger change beyond the scope of this PR. I do agree that this file is littered with lot of log tagging and requires some form of cleanup.

Comment (gow) [reply service/matching/matching_engine.go:869]: Resolving this to unblock the PR.

File: service/matching/config.go
Comment (yycptt) [inline service/matching/config.go:224]: nit: group the def and the assignment with other Query related config. I think there's another query related config above.

File: service/matching/matching_engine.go
Comment (yycptt) [inline service/matching/matching_engine.go:880]: nit: log runID as well.

**Review by yycptt — APPROVED** [review]

File: service/matching/config.go
Comment (yycptt) [reply service/matching/config.go]: +1, also filtering on task queue make sense. OSS users can still see the task queue tag in the metrics I think.

**Pull request merged**
